### ISSUE TITLE
NaN loss problem (related to nn.InstanceNorm1d(channels[i]))

### ISSUE THREAD

**Issue by GabbySuwichaya** [HeatherJiaZG/SuperGlue-pytorch#10]
Hello, @HeatherJiaZG.  Thank you for your effort to write these scripts. I respect your energy and time to make it publicly available.  

Here, I have a similar problem to other people about the Nan loss. 
From my observation (and my choice of local feature), the NaN loss came from MLP which seems to be associated with nn.InstanceNorm1d(channels[i]) in Line 58 in superglue.py.  

I am not sure if it is a good idea to turn it back to BatchNorm1D. 

At the moment, I am testing if BatchNorm1D would work. 
So, I would like to ask 3 main questions:

- why did you switch to nn.InstanceNorm1d ? Also, if you just use nn.InstanceNorm1d(x)...this mean that the parameters of  InstanceNorm1d is fixed (not trained). .... 

- Is that correct that the param of nn.InstanceNorm1d(x) should not be trained ? 

- And have you solve this problem?

**Comment by TangXiaoyu9** [comment]
Hi, did you solve the problem?

**Comment by GabbySuwichaya** [comment]
Hi @TangXiaoyu9. 
Just at the moment, I kind of solve the NaN, but I am not sure yet.  
Two changes that I have made are about 
1. Replacing  Relu() with PReLU() in [superglue](https://github.com/HeatherJiaZG/SuperGlue-pytorch/blob/master/models/superglue.py#L60) 
2. Add   **amsgrad=True**  in [Line172](https://github.com/HeatherJiaZG/SuperGlue-pytorch/blob/master/train.py#L172)

So, far it is still running for 22K iteration ....  Please share if you have a good solution ....

**Comment by TangXiaoyu9** [comment]
Thanks for your reply!

I tried clip_grad_norm(clip_value=0.1) and it is okay with a few thousand iterations so far, I am not sure if it is a good way to solve it though. Do let me know if you have any comments on this solution! :)

**Comment by GabbySuwichaya** [comment]
Hi @TangXiaoyu9... 
I have forgotten to let you know ..... 
  
1.  You should switch back to the BatchNorm1d in [Line58](https://github.com/HeatherJiaZG/SuperGlue-pytorch/blob/master/models/superglue.py#L58). 

I find that this also helps a lot with the convergence. I think many parts of this script was set up for SIFT keypoints. 

2.  Also, maybe you should reduce the **match_threshold** ... I have noticed that mine works at match_threshold  = 0.0... 
I know this is probably allowing bad matches, but I didn't get so many matches.  

3. At the moment, I am at about 55K iteration now. I am using the COCO training data.

**Comment by GabbySuwichaya** [comment]
I think I know the problem. The network architecture and other settings in this script are set for SIFT.   
So, using this script  can lead to NaN loss.

**Comment by atomishcv** [comment]
@GabbySuwichaya 
Hi, i use D2-Net to search [kps, soc, des] ,and use superglue to get inliner. when i training the code(after use your method to amend the code ), but i still meet the Loss NaN. cloud you give me some suggestion.

**Comment by GabbySuwichaya** [comment]
@atomishcv,
Actually, I have solved this problem, by redoing it from the official repo. I would recommend doing the following:
1. Use the model from the official site... Be careful that you will have to set the keypoint normalization for batch processing. 
2. Create your own ground truth annotation (matching labels).... 
3. I actually modify the loss function by looking at the focal loss from https://github.com/zju3dv/LoFTR (you need to look at the one with the Sinkhorn option). 

Hope this help and good luck.

**Comment by atomishcv** [comment]
@GabbySuwichaya thank you very much!
